### PR TITLE
[TD-2356]Add support for File sending events in Zendesk

### PR DESCRIPTION
**Pull request by raj5036**
Refer https://github.com/Kommunicate-io/Kommunicate-Server/pull/388

### CODE REVIEW

File: webplugin/js/app/zendesk-chat-service.js [hunk 1]
```diff
@@ -28,10 +28,30 @@ function ZendeskChatService() {
...
         window.console.log("handleUserMessage: ", event);
-        zChat.sendChatMsg(event.message.message, function (err, data) {
-            window.console.log("zChat.sendChatMsg ", err, data)
-        });
-        // TODO look for attachment
+        if(event.message.contentType==0){ //if user sends normal message
+            zChat.sendChatMsg(event.message.message, function (err, data) {
+                window.console.log("zChat.sendChatMsg ", err, data)
+            });
+            // TODO look for attachment
```
Comment (gargshantnu) [inline webplugin/js/app/zendesk-chat-service.js:35]: you can remove this :)

**Review by gargshantnu — DISMISSED** [review]

**Review by mukeshthawani — APPROVED** [review]

File: webplugin/js/app/zendesk-chat-service.js [hunk 2]
```diff
@@ -28,10 +28,29 @@ function ZendeskChatService() {
             return;
         }
         window.console.log("handleUserMessage: ", event);
-        zChat.sendChatMsg(event.message.message, function (err, data) {
-            window.console.log("zChat.sendChatMsg ", err, data)
-        });
-        // TODO look for attachment
+        if(event.message.contentType==0){ //if user sends normal message
+            zChat.sendChatMsg(event.message.message, function (err, data) {
+                window.console.log("zChat.sendChatMsg ", err, data)
```
Comment (Prochnost55) [inline webplugin/js/app/zendesk-chat-service.js:33]: is this contentType same as the contentType used in kommunicate's intent payload?

File: webplugin/js/app/zendesk-chat-service.js [hunk 3]
```diff
@@ -85,4 +106,33 @@ function ZendeskChatService() {
             },
         });
     };
+
+    _this.handleZendeskAgentFileSendEvent= async function (event) {
+        window.console.log("handleZendeskAgentFileSendEvent ",event);
+
+        // let {url, mime_type, name, size}=event.attachment; 
+        var messagePxy = {
```
Comment (Prochnost55) [inline webplugin/js/app/zendesk-chat-service.js:114]: is it working fine in IE11?

File: webplugin/js/app/zendesk-chat-service.js (hunk 2 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js]: Updated as requested.

File: webplugin/js/app/zendesk-chat-service.js [hunk 4]
```diff
@@ -85,4 +106,35 @@ function ZendeskChatService() {
             },
         });
     };
+
+    _this.handleZendeskAgentFileSendEvent= async function (event) {
+        // window.console.log("handleZendeskAgentFileSendEvent ",event);
+
+        // let {url, mime_type, name, size}=event.attachment; 
+        var messagePxy = {
```
Comment (Prochnost55) [inline webplugin/js/app/zendesk-chat-service.js:114]: We can remove async as we are not using await and comments if not required

File: webplugin/js/app/zendesk-chat-service.js [hunk 5]
```diff
@@ -28,10 +28,29 @@ function ZendeskChatService() {
...
-        });
-        // TODO look for attachment
+        if(event.message.contentType==KommunicateConstants.MESSAGE_CONTENT_TYPE.DEFAULT){ //if user sends normal message
+            zChat.sendChatMsg(event.message.message, function (err, data) {
+                window.console.log("zChat.sendChatMsg ", err, data)
+            });
+
+        }else{ //if user sends file attachments
+            let {name, thumbnailBlobKey, url, createdAtTime, contentType} = event.message.file;
+
```
Comment (Prochnost55) [inline webplugin/js/app/zendesk-chat-service.js:38]: the else part will get executed for location sharing also 
can you place a check?
Will switch case be better?

**Review by Prochnost55 — CHANGES_REQUESTED** [review]

File: webplugin/js/app/zendesk-chat-service.js (hunk 4 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js]: done

File: webplugin/js/app/zendesk-chat-service.js (hunk 5 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js]: Yes, used `else if` instead of `else`.

**Review by Prochnost55 — APPROVED** [review]

**Review by gargshantnu — APPROVED** [review]

**Review by mukeshthawani — DISMISSED** [review]

File: webplugin/js/app/zendesk-chat-service.js [hunk 6]
```diff
@@ -28,10 +28,26 @@ function ZendeskChatService() {
...
-        zChat.sendChatMsg(event.message.message, function (err, data) {
-            window.console.log("zChat.sendChatMsg ", err, data)
-        });
-        // TODO look for attachment
+        if(event.message.contentType == KommunicateConstants.MESSAGE_CONTENT_TYPE.DEFAULT){
+            zChat.sendChatMsg(event.message.message, function (err, data) {
+                window.console.log("zChat.sendChatMsg ", err, data)
+            });
+
+        }else if(event.message.contentType == KommunicateConstants.MESSAGE_CONTENT_TYPE.ATTACHMENT){
```
Comment (devashishmamgain) [inline webplugin/js/app/zendesk-chat-service.js:36]: please do the code formating

File: webplugin/js/app/zendesk-chat-service.js [hunk 7]
```diff
@@ -28,10 +28,26 @@ function ZendeskChatService() {
...
+
+            let fileInputElement=document.getElementById("mck-file-input");
```
Comment (devashishmamgain) [inline webplugin/js/app/zendesk-chat-service.js:38]: i guess let wouldn't work in chat widget code, @Prochnost55 @gargshantnu can you guys confirm?

File: webplugin/js/app/zendesk-chat-service.js [hunk 8]
```diff
@@ -85,4 +103,33 @@ function ZendeskChatService() {
             },
         });
     };
+
+    _this.handleZendeskAgentFileSendEvent= function (event) {
+        window.console.log("handleZendeskAgentFileSendEvent ",event);
+
+        var messagePxy = {
+            fileAttachment: event.attachment,
+            fromUserName: event.nick.split(":")[1],
```
Comment (devashishmamgain) [inline webplugin/js/app/zendesk-chat-service.js:112]: what is 'nick'?

File: webplugin/js/app/zendesk-chat-service.js [hunk 9]
```diff
@@ -85,4 +103,33 @@ function ZendeskChatService() {
...
+                data: JSON.stringify(messagePxy),
+                headers: {
+                    'x-authorization': window.Applozic.ALApiService.AUTH_TOKEN,
+                },
+                success: function (result) {
+                    console.log("Sent File message data to the server ", result);
+                    typeof callback == 'function' && callback(agentUserName);
+                },
+                error: function (err) {
+                    console.log('err while sending File message data to the server');
```
Comment (devashishmamgain) [inline webplugin/js/app/zendesk-chat-service.js:130]: some places it is window.console.log and some places it is console.log
which one is correct? and what's the purpose of using 2 different ways?

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/zendesk-chat-service.js (hunk 6 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js:36]: Okay

File: webplugin/js/app/zendesk-chat-service.js (hunk 9 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js:130]: Both are technically the same thing. As 'window' is global object for all browsers. 

Ok, Then I am using one format in all places (in this file).

File: webplugin/js/app/zendesk-chat-service.js (hunk 8 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js:112]: When a Zendesk agent sends a message, we get this 'nick' property in the event object. This 'nick' property contains their zendesk-ID, that we use in server side to fetch/store their details from the DB.
image

File: webplugin/js/app/zendesk-chat-service.js (hunk 7 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js:38]: 'let' is working, when running SDK locally.

Comment (devashishmamgain) [reply webplugin/js/app/zendesk-chat-service.js:38]: in IE? can you check it in opera mini as well
https://caniuse.com/?search=let

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

File: webplugin/js/app/zendesk-chat-service.js (hunk 7 above)
Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js:38]: Tested it in Opera, its working fine there 
In the below screenshot you can see in logs that "File has been sent successfully"
image

And as far IE is concerned, in my OS (Windows 11) IE is not supported. So have to ask Vipin to do that while he tests.

Comment (devashishmamgain) [reply webplugin/js/app/zendesk-chat-service.js:38]: why do we need to wait for Vipin on this? check https://www.lambdatest.com/ i think we have a subscription for it as well, check in QA channel

Comment (raj5036) [reply webplugin/js/app/zendesk-chat-service.js]: Yes it was giving an error for let keyword. So changed it to "var"

**Review by devashishmamgain — APPROVED** [review]

**Pull request merged**